src/load.c

```diff
@@ -362,6 +369,136 @@ int xmp_test_module(char *path, struct xmp_test_info *info)
 	return ret;
 }
 
+int xmp_test_module_from_memory(void *mem, long size, struct xmp_test_info *info)
+{
+	HIO_HANDLE *h;
+	char buf[XMP_NAME_SIZE];
+	int i;
+	int ret = -XMP_ERROR_FORMAT;
+
+	/* Use size < 0 for unknown/undetermined size */
+	if (size == 0)
+		size--;
+
+	if ((h = hio_open_mem(mem, size)) == NULL)
+		return -XMP_ERROR_SYSTEM;
+
+#ifndef LIBXMP_CORE_PLAYER
+	/* get size */
+	if (hio_size(h) < 256) {	/* set minimum valid module size */
+		ret = -XMP_ERROR_FORMAT;
+		goto err;
+	}
+#endif
+
+	if (info != NULL) {
+		*info->name = 0;	/* reset name prior to testing */
+		*info->type = 0;	/* reset type prior to testing */
+	}
+
+	for (i = 0; format_loader[i] != NULL; i++) {
+		hio_seek(h, 0, SEEK_SET);
+		if (format_loader[i]->test(h, buf, 0) == 0) {
+			int is_prowizard = 0;
+
+#ifndef LIBXMP_CORE_PLAYER
+			if (strcmp(format_loader[i]->name, "prowizard") == 0) {
+				hio_seek(h, 0, SEEK_SET);
+				pw_test_format(h, buf, 0, info);
+				is_prowizard = 1;
+			}
+#endif
+
+			hio_close(h);
+
+			if (info != NULL && !is_prowizard) {
+				strncpy(info->name, buf, XMP_NAME_SIZE - 1);
+				strncpy(info->type, format_loader[i]->name,
+							XMP_NAME_SIZE - 1);
+			}
+			return 0;
+		}
+	}
+
+#ifndef LIBXMP_CORE_PLAYER
+    err:
+	hio_close(h);
+#else
+	hio_close(h);
+#endif
+	return ret;
+}
+
+int xmp_test_module_from_file(void *file, long size, struct xmp_test_info *info)
+{
+	HIO_HANDLE *h;
+	char buf[XMP_NAME_SIZE];
+	int i;
+	int ret = -XMP_ERROR_FORMAT;
+#ifndef LIBXMP_CORE_PLAYER
+	char *temp = NULL;
+#endif
+	FILE *f = fdopen(fileno((FILE *)file), "rb");
```

Comment:
This line has the same bug as `xmp_load_module_from_file` that would be fixed by one of the patches in pull request #173.

Reply:
ohh, there is progress here again? This is awesome 👍 . Will take a look at the code later.

Reply:
For testing if a handle is supported it is imo not nice that afterwards my FILE-handle is trashed. I would expect that I can seek backwards and pass it directly to ``xmp_load_module_from_file``.
Though I won't really need the ``FILE`` case in my code. I only operate on memory buffers because the caller uses ``std::fstream`` which is not ``FILE``-compatible anyway.

Reply:
Currently, the progress here seems to be just sezero and I reviewing PRs...

Re: closing the `FILE` handle passed to `xmp_load_module_from_file` in that PR I linked, I agree that's not a great way to handle it. sezero posted a patch further down that PR that (IMO correctly) adds an hio function so that it can leave the user-provided `FILE` open without causing any portability issues with that `fdopen` hack.